src/Workspaces/Core/Portable/Workspace/Solution/DocumentInfo.cs

```diff
@@ -53,13 +54,21 @@ public sealed class DocumentInfo
         /// </summary>
         public TextLoader TextLoader { get; }
 
+        /// <summary>
+        /// A <see cref="IDocumentServiceFactory"/> associated with this document
+        /// this only exist in In-Proc and not part of roslyn solution checksum tree
+        /// in another word, for now, doesn't exist in OOP
+        /// </summary>
+        internal IDocumentServiceFactory DocumentServiceFactory { get; }
```

Comment:
having it on HostDocument might be enough. but in case we want to really generalize it so that things can be used in below VS layer, so passed it down. should be actually something like Document.GetService, but for now, here.